### PR TITLE
add rolestats command

### LINKED ISSUE #8: role statistics
for self-assignable (or all) roles, have a command to show stats on # of people with that role

### PR DESCRIPTION
**Pull request by classabbyamp**
fixes #8

### Example Usage
image
image
image

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -32,7 +32,11 @@
 intents.members = True
 intents.messages = True
 
-bot = commands.Bot(command_prefix=opt.command_prefix, intents=intents, member_cache_flags=discord.MemberCacheFlags.none(), chunk_guilds_at_startup=False)
+# Only cache members if needed by a bot feature
+member_cache_flags = discord.MemberCacheFlags.from_intents(intents) if opt.enable_role_stats else discord.MemberCacheFlags.none()
+chunk_guilds = opt.enable_role_stats
+
+bot = commands.Bot(command_prefix=opt.command_prefix, intents=intents, member_cache_flags=member_cache_flags, chunk_guilds_at_startup=chunk_guilds)
```
Comment (0x5c) [inline main.py]: Uselessly defined variable (`chunk_guilds`)

**Comment by classabbyamp** [comment]
Notes for review:
1. What should the check be for the `--all` subcommand be?
    - currently it is: `@commands.has_guild_permissions(administrator=True)`
2. Decided to abandon filtering to specific roles, is out of scope of the original issue and too complex to implement for now (discussion of this was the subject of the earlier confusion).
3. Might end up being too large for an embed. This seems unlikely, though (250 max # of roles, 2048 chars max in embed description). Possible solutions:
    - wait until `bolt` integration and use the embed limit checker
    - truncate it in this command now
4. Add any aliases for `rolestats`?
4. Add an alias to the `--all` subcommand for `all` (no dashes)?

File: selfrole.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 See LICENSE for the full text of the license.
 """
```
Comment (0x5c) [inline selfrole.py:10]: Missing line

File: selfrole.py [hunk 3]
```diff
@@ -137,3 +138,52 @@ async def apply_roles(self, ctx: commands.Context, roles: set):
                 embed.colour = 0xF04747
             embed.add_field(name="Invalid roles", value="\n".join(errors), inline=False)
         await ctx.send(embed=embed)
+
+    @commands.group(name="rolestats", enabled=opt.enable_role_stats)
+    @commands.guild_only()
+    async def role_stats(self, ctx: commands.Context):
+        """Show statistics for RoleMe-managed roles."""
```
Comment (0x5c) [inline selfrole.py]: Capitalising the command name makes no sense if the command itself is not capitalised. This capitalisation exists nowhere else in the bot.

**Comment by classabbyamp** [comment]
> Notes for review:
> 
> 1. What should the check be for the `--all` subcommand be?
>    * currently it is: `@commands.has_guild_permissions(administrator=True)`

ok

> 2. Decided to abandon filtering to specific roles, is out of scope of the original issue and too complex to implement for now (discussion of this was the subject of the earlier confusion).

ACK

> 3. Might end up being too large for an embed. This seems unlikely, though (250 max # of roles, 2048 chars max in embed description). Possible solutions:
>    * wait until `bolt` integration and use the embed limit checker
>    * truncate it in this command now

4 elements (name, mention, id, count). <25: field per role, >=25: in desc, cont'd in fields (limited to embed limits)

> 4. Add any aliases for `rolestats`?

no clue

> 5. Add an alias to the `--all` subcommand for `all` (no dashes)?

yes

(from @0x5c)

File: selfrole.py [hunk 4]
```diff
@@ -137,3 +138,52 @@ async def apply_roles(self, ctx: commands.Context, roles: set):
...
+            role_stats: List[Tuple[str, int]] = []
+
+            if ctx.guild.id not in opt.selfroles:
+                # There is no selfrole configured for this guild.
+                embed = discord.Embed(title="Error!", colour=0xF04747)
+                embed.description = "This guild is not configured for self-assignable roles! RoleMe statistics are not available."
+                await ctx.send(embed=embed)
+                return
+
+            roles = [ctx.guild.get_role(r) for r in opt.selfroles[ctx.guild.id]]
```
Comment (0x5c) [inline selfrole.py:158]: What if a role in the config is invalid and returns `None`?

**Review by 0x5c — CHANGES_REQUESTED** [review]

**Comment by classabbyamp** [comment]
### New Interface

image

image

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,63 +1,64 @@
...
-
-[Unreleased]: https://github.com/0x5c/minibot/compare/v1.3.2...HEAD
-[1.3.2]: https://github.com/0x5c/minibot/releases/tag/v1.3.2
-[1.3.1]: https://github.com/0x5c/minibot/releases/tag/v1.3.1
-[1.3.0]: https://github.com/0x5c/minibot/releases/tag/v1.3.0
-[1.2.1]: https://github.com/0x5c/minibot/releases/tag/v1.2.1
-[1.2.0]: https://github.com/0x5c/minibot/releases/tag/v1.2.0
-[1.0.1]: https://github.com/0x5c/minibot/releases/tag/v1.0.1
-[1.0.0]: https://github.com/0x5c/minibot/releases/tag/v1.0.0
+# Changelog for Minibot
```
Comment (0x5c) [inline CHANGELOG.md]: Woweeee
Why is this entire file rendering as changed? 😨

File: selfrole.py [hunk 6]
```diff
@@ -137,3 +139,90 @@ async def apply_roles(self, ctx: commands.Context, roles: set):
...
         await ctx.send(embed=embed)
+
+    @commands.group(name="rolestats", enabled=opt.enable_role_stats)
+    @commands.guild_only()
+    async def _role_stats(self, ctx: commands.Context):
+        """Show statistics for roleme-managed roles."""
+        if ctx.invoked_subcommand is None:
+            role_stats: Dict[discord.Role, int] = {}
+
+            if ctx.guild:
```
Comment (0x5c) [inline selfrole.py]: What is that one solving?
Getting the guild should always be possible, unless the cache settings are not correct.

File: selfrole.py [hunk 7]
```diff
@@ -137,3 +139,90 @@ async def apply_roles(self, ctx: commands.Context, roles: set):
...
+                if ctx.guild.id not in opt.selfroles:
+                    # There is no selfrole configured for this guild.
+                    embed = discord.Embed(title="Error!", colour=0xF04747)
+                    embed.description = ("This guild is not configured for self-assignable roles! "
```
Comment (0x5c) [inline selfrole.py]: Maybe add a newline?

File: selfrole.py [hunk 8]
```diff
@@ -137,3 +139,90 @@ async def apply_roles(self, ctx: commands.Context, roles: set):
...
+                await ctx.send(embed=embed)
+
+    @_role_stats.command(name="--all", aliases=["all", "-a", "a"], enabled=opt.enable_role_stats)
+    @commands.guild_only()
+    @commands.has_guild_permissions(administrator=True)
+    async def role_stats_all(self, ctx: commands.Context):
+        """Show statistics for all roles. This command is admin-only."""
+        role_stats: Dict[discord.Role, int] = {}
+
+        if ctx.guild:
```
Comment (0x5c) [inline selfrole.py]: Same as for the other

**Review by 0x5c — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 5 above)
Comment (classabbyamp) [reply CHANGELOG.md]: ya no clue

File: selfrole.py (hunk 6 above)
Comment (classabbyamp) [reply selfrole.py]: mypy yells because `ctx.guild` is technically optional

File: selfrole.py (hunk 8 above)
Comment (classabbyamp) [reply selfrole.py]: Same as for the other

File: selfrole.py (hunk 6 above)
Comment (0x5c) [reply selfrole.py]: This is a case of "shut-up mypy"
Mypy is not run on the repo yet, so let's ignore it.

Later we'll have to find how to tell it to shut up when the condition is impossible.

File: selfrole.py (hunk 7 above)
Comment (0x5c) [reply selfrole.py]: (in the string)

File: CHANGELOG.md (hunk 5 above)
Comment (0x5c) [reply CHANGELOG.md]: What are the line endings?

Comment (0x5c) [reply CHANGELOG.md]: Will require local restoration of the history

**Review by 0x5c — APPROVED** [review]

**Pull request merged**